src/java/org/apache/cassandra/hints/HintsDispatcher.java

```diff
@@ -165,10 +226,62 @@ private Action sendHintsAndAwait(HintsReader.Page page)
         else
         {
             HintDiagnostics.pageSuccessResult(this, success, failures, timeouts);
+            rehintHintsNeedingRehinting();
             return Action.CONTINUE;
         }
     }
 
+    private void rehintHintsNeedingRehinting()
+    {
+        ClusterMetadata cm = ClusterMetadata.current();
+        Hint hint;
+        while ((hint = hintsNeedingRehinting.poll()) != null)
+        {
+            HintsService.instance.writeForAllReplicas(hint);
+            Mutation mutation = hint.mutation;
+            // Also may need to apply locally because it's possible this is from the batchlog
+            // and we never applied it locally
+            // TODO (review): Additional error handling necessary? Hints are lossy
```

Comment:
agreed, they're best effort

Reply:
Except they are how the batchlog does retries and we go to surprising lengths to preserve hints. I recall seeing that we stream hints to replacement nodes.